programs/perpetuals/src/state/custody.rs

```diff
@@ -181,6 +211,62 @@ impl Custody {
             && self.oracle.validate()
             && self.pricing.validate()
             && self.fees.validate()
+            && self.borrow_rate.validate()
+    }
+
+    pub fn update_borrow_rate(&mut self, curtime: i64) -> Result<()> {
+        // if current_utilization < optimal_utilization:
+        //   rate = base_rate + (current_utilization / optimal_utilization) * slope1
+        // else:
+        //   rate = base_rate + slope1 + (current_utilization - optimal_utilization) / (1 - optimal_utilization) * slope2
+
+        if curtime <= self.borrow_rate_state.last_update {
+            return Ok(());
+        }
+
+        let current_utilization = math::checked_div(
+            math::checked_mul(self.assets.locked as u128, Perpetuals::RATE_POWER)?,
+            self.assets.owned as u128,
+        )?;
+
+        let hourly_rate = if current_utilization < (self.borrow_rate.optimal_utilization as u128)
+            || (self.borrow_rate.optimal_utilization as u128) >= Perpetuals::RATE_POWER
+        {
+            math::checked_div(
+                math::checked_mul(current_utilization, self.borrow_rate.slope1 as u128)?,
+                self.borrow_rate.optimal_utilization as u128,
+            )?
+        } else {
+            math::checked_add(
+                self.borrow_rate.slope1 as u128,
+                math::checked_div(
+                    math::checked_mul(
+                        math::checked_sub(
+                            current_utilization,
+                            self.borrow_rate.optimal_utilization as u128,
+                        )?,
+                        self.borrow_rate.slope2 as u128,
+                    )?,
+                    Perpetuals::RATE_POWER - self.borrow_rate.optimal_utilization as u128,
+                )?,
+            )?
+        };
+        let hourly_rate = math::checked_add(
+            math::checked_as_u64(hourly_rate)?,
+            self.borrow_rate.base_rate,
+        )?;
+
+        let rate_per_second = math::checked_div(hourly_rate, 3600)?;
+        let rate_sum = math::checked_mul(
```

Comment:
so the `rate_sum` is just the compounded % rate since last update. What would that be used for?

Reply:
not really, if you divide first you lose the precision, because this is an integer division (i.e. result is rounded). 
for example:
12345 / 10000 * 10000 = 10000
but
12345 * 10000 / 10000 = 12345
to ensure multiplication doesn't overflow I cast to u128 first

Reply:
not really a compounded, but rather a sum, so user will pay a simple interest rate at position close or liquidation. I'm thinking if I should make it compounded. In such case accumulated interest will grow exponentially, discouraging users to hold position for too long. But maybe this is what we want.

Reply:
I would imagine compounded interest rates being sub-optimal.

Reply:
please check out the updated code and let me know if there are any concerns!